File: src/main/python/AoC2021_08.py
#! /usr/bin/env python3
#
# Advent of Code 2021 Day 8
#

from __future__ import annotations

import sys
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST = """\
be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | \
fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | \
fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | \
cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | \
efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | \
gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | \
gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | \
cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | \
ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | \
gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | \
fgae cfgab fg bagce
"""


class SignalAndOutput(NamedTuple):
    signals: list[str]
    outputs: list[str]

    @classmethod
    def from_str(cls, s: str) -> SignalAndOutput:
        signals = ["".join(sorted(_)) for _ in s.split(" | ")[0].split()]
        outputs = ["".join(sorted(_)) for _ in s.split(" | ")[1].split()]
        return SignalAndOutput(signals, outputs)


Input = list[SignalAndOutput]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [SignalAndOutput.from_str(s) for s in input_data]

    def part_1(self, input: Input) -> int:
        return sum(
            len(word) in {2, 3, 4, 7}
            for line in (line for _, line in input)
            for word in line
        )

    def part_2(self, input: Input) -> int:
        def solve(input: SignalAndOutput) -> int:
            def find(segments: int, expected: int) -> set[str]:
                ans = {_ for _ in input.signals if len(_) == segments}
                assert len(ans) == expected
                return ans

            def shares_all_segments(container: str, contained: str) -> bool:
                cd = set(contained)
                return set(container) & cd == cd

            digits = dict[str, str]()
            # 1
            twos = find(2, 1)
            one = list(twos)[0]
            digits[one] = "1"
            # 7
            threes = find(3, 1)
            digits[list(threes)[0]] = "7"
            # 4
            fours = find(4, 1)
            four = list(fours)[0]
            digits[four] = "4"
            # 8
            sevens = find(7, 1)
            digits[list(sevens)[0]] = "8"
            # 9
            sixes = find(6, 3)
            possible_nines = {
                six for six in sixes if shares_all_segments(six, four)
            }
            assert len(possible_nines) == 1
            nine = list(possible_nines)[0]
            digits[nine] = "9"
            # 0
            possible_zeroes = {
                six
                for six in sixes
                if six != nine and shares_all_segments(six, one)
            }
            assert len(possible_zeroes) == 1
            zero = list(possible_zeroes)[0]
            digits[zero] = "0"
            # 6
            possible_sixes = {
                six for six in sixes if six != nine and six != zero
            }
            assert len(possible_sixes) == 1
            digits[list(possible_sixes)[0]] = "6"
            # 3
            fives = find(5, 3)
            possible_threes = {
                five for five in fives if shares_all_segments(five, one)
            }
            assert len(possible_threes) == 1
            three = list(possible_threes)[0]
            digits[three] = "3"
            # 5
            possible_fives = {
                five
                for five in fives
                if five != three and shares_all_segments(nine, five)
            }
            assert len(possible_fives) == 1
            five = list(possible_fives)[0]
            digits[five] = "5"
            # 2
            possible_twos = {f for f in fives if f != five and f != three}
            assert len(possible_twos) == 1
            digits[list(possible_twos)[0]] = "2"

            return int("".join(digits[o] for o in input.outputs))

        return sum(solve(_) for _ in input)

    @aoc_samples(
        (
            ("part_1", TEST, 26),
            ("part_2", TEST, 61229),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2021, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/julia/AoC2021_08.jl
#! /usr/bin/env julia
#

include("aoc.jl")
include("aocd.jl")
using .Aocd

function _parse(input)
    return ([[join(sort(collect(x)))
              for x ∈ split(split(line, " | ")[i])]
             for line ∈ input]
            for i ∈ 1:2)
end

function part1(input)
    signals, outputs = _parse(input)
    return sum(length(word) ∈ [2, 3, 4, 7]
               for line in outputs
               for word in line)
end

function part2(input)
    function solve(signals, outputs)
        function find(signals, segments, expected)
            ans = filter(x -> length(x) == segments, signals)
            @assert length(ans) == expected
            return ans
        end
        function shares_all_segments(container, contained)
            return all(∈(container), contained)
        end

        digits = Dict{String,Char}()
        # 1
        twos = find(signals, 2, 1)
        one = twos[1]
        digits[one] = '1'
        # 7
        threes = find(signals, 3, 1)
        digits[threes[1]] = '7'
        # 4
        fours = find(signals, 4, 1)
        four = fours[1]
        digits[four] = '4'
        # 8
        sevens = find(signals, 7, 1)
        digits[sevens[1]] = '8'
        # 9
        sixes = find(signals, 6, 3)
        possible_nines = filter(x -> shares_all_segments(x, four), sixes)
        @assert length(possible_nines) == 1
        nine = possible_nines[1]
        digits[nine] = '9'
        # 0
        possible_zeroes = filter(x -> x != nine && shares_all_segments(x, one),
                                 sixes)
        @assert length(possible_zeroes) == 1
        zero = possible_zeroes[1]
        digits[zero] = '0'
        # 6
        possible_sixes = filter(x -> x != nine && x != zero, sixes)
        @assert length(possible_sixes) == 1
        digits[possible_sixes[1]] = '6'
        # 3
        fives = find(signals, 5, 3)
        possible_threes = filter(x -> shares_all_segments(x, one), fives)
        @assert length(possible_threes) == 1
        three = possible_threes[1]
        digits[three] = '3'
        # 5
        possible_fives = filter(x -> x != three && shares_all_segments(nine, x),
                                fives)
        @assert length(possible_fives) == 1
        five = possible_fives[1]
        digits[five] = '5'
        # 2
        possible_twos = filter(x -> x != five && x != three, fives)
        @assert length(possible_twos) == 1
        digits[possible_twos[1]] = '2'

        return parse(Int, join(digits[o] for o ∈ outputs))
    end

    signals, outputs = _parse(input)
    return sum(solve(signals[i], outputs[i])
               for i ∈ 1:length(input))
end

TEST = @aoc_splitlines("""
be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | \
fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | \
fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | \
cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | \
efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | \
gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | \
gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | \
cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | \
ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | \
gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | \
fgae cfgab fg bagce
""")

function samples()
    @assert part1(TEST) == 26
    @assert part2(TEST) == 61_229
end

@aoc_main 2021 8
